src/toil/test/src/jobCacheTest.py

```diff
@@ -1125,6 +1126,154 @@ def testBatchSystemCleanupCanHandleWorkerDeaths(self):
             assert not os.path.exists(nonLocalFile1)
             assert not os.path.exists(nonLocalFile2)
 
+        # Test to check if files created written to the jobstore on failed/killed jobs are deleted
+        # or not.
+        def testOrphanedFilesAreDeletedFromJobStoreOnKilledJob(self):
+            """
+            The "A-is-killed" version of the orphaned file tests.
+            """
+            self._abstractTestOrphanedFilesAreDeletedFromJobStore('kill')
+
+        def testOrphanedFilesAreDeletedFromJobStoreOnFailedJob(self):
+            """
+            The "A-failed" version of the orphaned file tests.
+            """
+            self._abstractTestOrphanedFilesAreDeletedFromJobStore('fail')
+
+        def _abstractTestOrphanedFilesAreDeletedFromJobStore(self, testCase):
+            """
+            Run a job that creates a file, writes it to the file store and records the jobstore file
+            ID to an external file before dying/being killed.  Ensure that the file is deleted from
+            the jobstore at the appropriate time.
+
+            :param str testCase: Is this the `A-is-killed` test or the `A-failed` test?
+            """
+            assert testCase in ('kill', 'fail')
+            # There can be no retries
+            self.options.retryCount = 0
+            self.options.logLevel = 'DEBUG'
+            workdir = self._createTempDir(purpose='nonLocalDir')
+            nonLocalFile = os.path.join(workdir, str(uuid4()))
+            # nonLocalFile has to exist and be empty
+            open(nonLocalFile, 'w').close()
+            assert os.path.exists(nonLocalFile)
+            root = Job()
+            A = Job.wrapJobFn(self._testOrphanedFilesAreDeletedFromJobStore_A,
```

Comment:
Not necessarily for this PR but I thought of a naming convention for jobs: letters for sibliings (left to right in the tree) numbers for descendants. So here it would be A, B1 and B2. If B1 had two children they would be B2A and B2B.